### PR TITLE
`ClExprOp` and friends

### PR DESCRIPTION
**Pull request by cqc-alec**
This is the first stage of `ClassicalExpBox` refactoring: defining a new operation type (`ClExprOp`) to replace it. This is defined in C++ with Python bindings, and is independent of bit and register names (information about which variable corresponds to which input etc. is encoded in the `Op`).

This PR also implements QASM conversion to and from the new op, but for QASM-to-circuit conversion this is not the default: a new boolean argument for `circuit_from_qasm` and friends is added to control this. This is so as not to break current workflows that rely on pytket-to-X conversions that don't currently handle the `ClExprOp`. After this is in a release we will update pytket-qir, pytket-phir and any other relevant extensions to handle the new type; after that we can update tket to use it everywhere by default, write a new decomposer pass for it, and deprecate `ClassicalExpBox`.

PS The Nix build failure is expected because it is using an older mypy version.

### CODE REVIEW

**Review by sjdilkes — CHANGES_REQUESTED** [review]
I've only left a couple of comments - looks nice and clean. Could it be worth adding tests covering some of the strange behaviour from `ClassicalExpBox`, such as having classical expressions in `CircBox` that are copied, or python circuits that have classical expressions added, the units relabelled and then converted to qasm to confirm that the expressions are as desired?

File: tket/include/tket/Ops/ClExpr.hpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+#include <ostream>
+#include <set>
+#include <variant>
+#include <vector>
+
+#include "tket/Ops/Op.hpp"
+
+namespace tket {
+
+// TODO Use X or list macros to reduce boilerplate.
```
[the remark this answers is not in the document]
Comment (cqc-alec) [reply tket/include/tket/Ops/ClExpr.hpp:33]: Not really a concern, but the code could be made more compact; I'm happy to leave this for a future PR though.

**Comment by cqc-alec** [comment]
> I've only left a couple of comments - looks nice and clean. Could it be worth adding tests covering some of the strange behaviour from `ClassicalExpBox`, such as having classical expressions in `CircBox` that are copied, or python circuits that have classical expressions added, the units relabelled and then converted to qasm to confirm that the expressions are as desired?

Good idea, I'll add some such tests.

File: tket/include/tket/Ops/ClExpr.hpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+  /**
+   * Default constructor
+   */
+  WiredClExpr();
+
+  /**
+   * Construct by specifying the bit, register and output positions
+   *
+   * @param expr Expression
+   * @param bit_posn Map from identifiers of bit variables to bit positions
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/include/tket/Ops/ClExpr.hpp:224]: Thanks!

File: tket/include/tket/Ops/ClExpr.hpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+              {ClOp::RegDiv, "RegDiv"},   {ClOp::RegPow, "RegPow"},
+              {ClOp::RegLsh, "RegLsh"},   {ClOp::RegRsh, "RegRsh"},
+              {ClOp::RegNeg, "RegNeg"},
+          })
+
+/**
+ * A bit variable within an expression
+ */
+typedef struct ClBitVar {
+  unsigned i;  /// Identifier for the variable within the expression
```
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/include/tket/Ops/ClExpr.hpp]: Sounds good, thanks.

**Review by sjdilkes — APPROVED** [review]
Exciting, LGTM! So I could submit jobs to `QuantinuumBackend` with `qasm` conversion and `ClExpr` now and it should work?

**Comment by cqc-alec** [comment]
> Exciting, LGTM! So I could submit jobs to `QuantinuumBackend` with `qasm` conversion and `ClExpr` now and it should work?

Yes I believe that should work. Note that there is not yet a version of `DecomposeClassicalExp` that works with `ClExpr`, but that should not be needed in many cases.

**Pull request merged**